### PR TITLE
SSmetrics

### PR DESCRIPTION
**Pull request by AffectedArc07**
## About The Pull Request

:wave:

Ports https://github.com/ParadiseSS13/Paradise/pull/16549

TLDR: Lets crossed (and whoever else has kibana access) look at server stats on graphs and use this data for meaningful things. See original PR for more detail

### Note

I didnt add SSair metrics because I have no idea what metrics auxmos needs. Tell me in code review if it matters

## Why It's Good For The Game

See above.

## Changelog
:cl: AffectedArc07
server: Added SSmetrics. This has config changes. 
/:cl:

### CODE REVIEW

**Comment by ike709** [comment]
Have you profiled the performance hit of this?

**Comment by AffectedArc07** [comment]
> Have you profiled the performance hit of this?

image

Negligible

**Comment by jupyterkat** [comment]
because of how auxmos works every turf is active. there's just `high_pressure_turfs` and `low_pressure_turfs`, which are turfs being actually processed

File: code/controllers/subsystem/metrics.dm
Comment (Crossedfall) [inline code/controllers/subsystem/metrics.dm]: Already talked to Affected about this over Discord. We'll need a new proc that provides an ISO friendly timestamp before it can be properly test merged. We could either modify SQLtime to have an optional Boolean to output a properly compliant ISO timestamp, or, we could have a new proc. I don't personally care either way, one of the other maintainers might though. This would be my recommendation.

```dm
/proc/iso_timestamp(timevar)
    return time2text(timevar || world.timeofday, "YYYY-MM-DDThh:mm:ss")
```

**Review by Crossedfall — CHANGES_REQUESTED** [review]
Everything looks fine based on my initial tests and review.  Only major problem identified was with the timestamp.

image

File: code/controllers/subsystem/metrics.dm
Comment (AffectedArc07) [reply code/controllers/subsystem/metrics.dm]: `get_metrics_json returned: {"@timestamp":"2021-08-21T12:10:47","cpu":1.07202, ...`

Done

File: code/__DEFINES/MC.dm
Comment (ike709) [inline code/__DEFINES/MC.dm]:
```suggestion
	ss_id="timer_[#X]";\
```

File: code/controllers/subsystem.dm
Comment (ike709) [inline code/controllers/subsystem.dm:5]: You currently only set this on the timer and processing subsystems, not regular subsystems.

File: code/__DEFINES/MC.dm
Comment (AffectedArc07) [reply code/__DEFINES/MC.dm]: Oh shit I forgot the real SS one

File: code/controllers/subsystem/metrics.dm
Comment (ike709) [inline code/controllers/subsystem/metrics.dm]:
```suggestion
/client/verb/debugmetrics()
```

File: code/modules/unit_tests/subsystem_metric_sanity.dm
Comment (ike709) [inline code/modules/unit_tests/subsystem_metric_sanity.dm:3]: You should validate `ss_id` in the unit test :^)

**Comment by Crossedfall** [comment]
TM confirms functionality within our infra. 👍

File: code/controllers/subsystem/metrics.dm
Comment (Crossedfall) [inline code/controllers/subsystem/metrics.dm:42]: Wasn't something I thought about until after the TM. We'll need an additional field that tells us which server the data is actually coming from. Round IDs are unique, but they don't give an at-a-glance understanding of which server it happened on.

```suggestion
	out["round_id"] = text2num(GLOB.round_id) // This is so we can filter the metrics by a single round ID
	out["server_name"] = CONFIG_GET(string/serversqlname) // This is so we can filter by a single server
```

Comment (Crossedfall) [inline code/controllers/subsystem/metrics.dm:42]: Wasn't something I thought about until after the TM. We'll need an additional field that tells us which server the data is actually coming from. Round IDs are unique, but they don't give an at-a-glance understanding of which server it happened on.

```suggestion
	var/list/out = list()
	var/server_name = CONFIG_GET(string/serversqlname)
	out["@timestamp"] = iso_timestamp() // This is required by ElasticSearch, complete with this name. DO NOT REMOVE THIS.
	out["cpu"] = world.cpu
	out["maptick"] = world.map_cpu
	out["elapsed_processed"] = world.time
	out["elapsed_real"] = (REALTIMEOFDAY - world_init_time)
	out["client_count"] = length(GLOB.clients)
	out["round_id"] = text2num(GLOB.round_id) // This is so we can filter the metrics by a single round ID
	if(server_name)
		out["server_name"] = server_name  // This is so we can filter by a single server
```

Comment (Crossedfall) [reply code/controllers/subsystem/metrics.dm:42]: I hate how the suggestions thing tabs 😩

**Review by ike709 — APPROVED** [review]

**Review by Crossedfall — APPROVED** [review]

**Comment by Crossedfall** [comment]
Thanks for porting that for us ❤️

**Pull request merged**
